docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,1063 @@
+}
+```
+
+**Validation and defaulting rules for Provider and ProviderSpec**
+- The `Name` field within `metav1.ObjectMeta` could be any valid Kubernetes
+  name; however, it is recommended to use Cluster API provider names. For
+  example, aws, vsphere, kubeadm. These names will be used to fetch the
+  default configurations in case there is no specific FetchConfiguration
+  defined.
+- `ProviderSpec.Version` should be a valid default version with the "v" prefix
+  as commonly used in the Kubernetes ecosystem; if this value is nil when a
+  new provider is created, the operator will determine the version to use
+  applying the same rules implemented in clusterctl (latest).
+  Once the latest version is calculated it will be set in
+  `ProviderSpec.Version`.
+- The content of `ProviderSpec.SecretName` will be treated as immutable. If
+  changes need to be made, a new object can be created and the name should be
+  updated.
+- The content of `ProviderSpec.SecretName` should be in the form of key:value.
+- The `ProviderSpec.SecretName` specified must be in the same namespace as the
+  provider.
+
+**Validation rules for ProviderSpec.FetchConfiguration**
+- If the FetchConfiguration is empty and not defined, then the operator will
+  apply the embedded fetch configuration for the given kind and
+  `ObjectMeta.Name`. For example, the infrastructure name `aws` will fetch
+  artifacts from
+  https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases.
+- If FetchConfiguration is not nil, exactly one of `URL` or `Selector` must be
+  specified.
+- `FetchConfiguration.Selector` is used to fetch provider’s components and
+  metadata from ConfigMaps stored inside the cluster. Each ConfigMap is
+  expected to contain components and metadata for a specific version only. So
+  if multiple versions of the providers need to be specified, they can be
+  added as separate ConfigMaps and labeled with the same selector. This
+  provides the same behavior as the “local” provider repositories but now from
+  within the management cluster.
+- `FetchConfiguration` is used only during init and upgrade operations.
+  Changes made to the contents of `FetchConfiguration` will not trigger a
+  reconciliation. This is similar behavior to `ProviderSpec.SecretName`.
+
+**Validation Rules for ProviderSpec.ManagerSpec**
+- The ControllerManagerConfigurationSpec is a type from
+  `controller-runtime/pkg/config` and is an embedded into the `ManagerSpec`.
+  This type will expose LeaderElection, SyncPeriod, Webhook, Health and
+  Metrics configurations.
+- If `ManagerSpec.Debug` is set to true, the operator will not allow changes
+  to other properties since it is in Debug mode.
+- If you need to set specific concurrency values for each reconcile loop (e.g.
+  `awscluster-concurrency`), you can leave
+  `ManagerSpec.MaxConcurrentReconciles` nil and use `Container.Args`.
+- If `ManagerSpec.MaxConcurrentReconciles` is set and a specific concurrency
```

Comment:
I like this logic, `MaxConcurrentReconciles` is the overall concurrency but any specific flag tuning takes precedence

Reply:
Yup. As mentioned above, I suggest this be deferred to the individual controller level. That is, if the user has set `MaxConcurrentReconciles` and also defined a specific flag, then the operator can pass to the controller `--controller-concurrency=20 --awscluster-concurrency=10` and the controller manager can override `awscluster-concurrency` accordingly.

Apologies if this is getting too "implementationy".